### PR TITLE
Simplify more expressions in `SimplifyingIrBuilder`

### PR DESCRIPTION
**Pull request by jacobhinkle**
This PR adds handling methods for comparison operators in `SimplifyingIrBuilder`. It also adds `sameAs` checks to a few existing methods to simplify expressions like `s0 - s0 = 0`, `ceilDiv(s1, s1) = 1`, or `s2 % s2 = 0`.

To implement the simplifying comparisons, we introduce `csrc/C++20/compare` holding `partial_ordering` which is a class that can represent a three-way comparison with an "unordered" possibility. In this case, if either of the arguments to `SimplifyingIrBuilder::ltExpr(Val* lhs, Val* rhs)` are non-constant, the comparison is unordered and we return `IrBuilder::ltExpr(lhs, rhs)`.

### CODE REVIEW

**Comment by jacobhinkle** [comment]
!build

**Comment by jacobhinkle** [comment]
!build

**Comment by jacobhinkle** [comment]
This does some simplification in the binary test suite. For example in `CatOfBroadcast`:
```diff
--- codegen_comparison/7d539671/custom_command_20230918_144131/cuda/__tmp_kernel3335.cu 2023-09-18 15:06:42.268298215 -0400
+++ codegen_comparison/93137140/custom_command_20230918_144131/cuda/__tmp_kernel3335.cu 2023-09-18 15:32:58.830221927 -0400
@@ -9617,21 +9617,21 @@
   }
   NVFUSER_UPDATE_MAGIC_ZERO;
   float T2[8];
   #pragma unroll
   for(nvfuser_index_t i7 = 0; i7 < 4; ++i7) {
     nvfuser_index_t i8;
     i8 = 2 * i7;
     #pragma unroll
     for(nvfuser_index_t i9 = 0; i9 < 2; ++i9) {
       T2[(i8 + i9)]
-         = true ? T0[(T0.alloc_stride[1] * (i9 + nvfuser_zero))] : 0.000000000e+00f;
+         = ((0 >= 0) && (0 < 1)) ? T0[(T0.alloc_stride[1] * (i9 + nvfuser_zero))] : 0.000000000e+00f;
     }
   }
   NVFUSER_UPDATE_MAGIC_ZERO;
```
We do not seem to simplify the `where` expression away in this case. This is because `IndexLowering::handle(const PadOp*)` manually creates a `TernaryOp` instead of using `SimplifyingIrBuilder`.

File: csrc/ir/builder.cpp [hunk 1]
```diff
@@ -330,6 +332,9 @@ Val* SimplifyingIrBuilder::addExpr(Val* lhs, Val* rhs) {
 }
 
 Val* SimplifyingIrBuilder::subExpr(Val* lhs, Val* rhs) {
+  if (lhs->sameAs(rhs)) {
```
Comment (zasdfgbnm) [inline csrc/ir/builder.cpp:335]: What if `lhs` is NAN?

File: csrc/ir/builder.cpp [hunk 2]
```diff
@@ -373,12 +378,17 @@ Val* SimplifyingIrBuilder::divExpr(Val* lhs, Val* rhs) {
   if (rhs->isOneInt()) {
     return lhs;
   }
+  if (lhs->sameAs(rhs)) {
```
Comment (zasdfgbnm) [inline csrc/ir/builder.cpp:381]: What if `rhs` is zero or inf or nan?

File: csrc/ir/builder.cpp [hunk 3]
```diff
@@ -373,12 +378,17 @@ Val* SimplifyingIrBuilder::divExpr(Val* lhs, Val* rhs) {
...
+    return lhs->fusion()->oneVal(lhs->dtype());
+  }
   return IrBuilder::divExpr(lhs, rhs);
 }
 
 Val* SimplifyingIrBuilder::ceilDivExpr(Val* lhs, Val* rhs) {
   if (rhs->isOneInt()) {
     return lhs;
+  } else if (lhs->sameAs(rhs)) {
```
Comment (zasdfgbnm) [inline csrc/ir/builder.cpp:390]: What if rhs is zero or inf or nan?

File: csrc/ir/builder.cpp [hunk 4]
```diff
@@ -391,7 +401,7 @@ Val* SimplifyingIrBuilder::ceilDivExpr(Val* lhs, Val* rhs) {
 }
 
 Val* SimplifyingIrBuilder::modExpr(Val* lhs, Val* rhs) {
-  if (rhs->isOneInt()) {
+  if (rhs->isOneInt() || lhs->isZeroInt() || lhs->sameAs(rhs)) {
```
Comment (zasdfgbnm) [inline csrc/ir/builder.cpp:404]: Should we `NVFUSER_ERROR` that both `lhs` and `rhs` are integral types?

File: csrc/ir/builder.cpp [hunk 5]
```diff
@@ -600,25 +612,118 @@ Val* SimplifyingIrBuilder::gcdExpr(Val* lhs, Val* rhs) {
   if (rhs->isZeroInt()) {
     return lhs;
```
Comment (zasdfgbnm) [inline csrc/ir/builder.cpp:613]: Could you also add an `NVFUSER_ERROR` check that inputs are integral types?

File: csrc/ir/builder.cpp (hunk 1 above)
Comment (jacobhinkle) [reply csrc/ir/builder.cpp:335]: Ah good point. I can guard this with `lhs->isIntegralScalar()` then it's valid right?

File: csrc/ir/builder.cpp (hunk 4 above)
Comment (jacobhinkle) [reply csrc/ir/builder.cpp:404]: I think I can just guard the last condition here with `lhs->isIntegralScalar()`.

**Comment by jacobhinkle** [comment]
Question for you @zasdfgbnm regarding safety of these simplifcations: some of the other `SimplifyingIrBuilder` methods check for nullptrs, but not all do. When would we expect to have null inputs?

File: csrc/ir/builder.cpp (hunk 2 above)
Comment (jacobhinkle) [reply csrc/ir/builder.cpp:381]: Yeah, I think I need to remove this.

File: csrc/ir/builder.cpp (hunk 1 above)
Comment (zasdfgbnm) [reply csrc/ir/builder.cpp:335]: I think so

File: csrc/ir/builder.cpp (hunk 3 above)
Comment (jacobhinkle) [reply csrc/ir/builder.cpp:390]: Guarding with `isIntegralScalar` handles inf and nan, but I don't have a solution for zero. Leaving as is, we might replace `ceilDiv(0, 0)` which should raise an exception with `1`.  I will remove this branch.

File: csrc/ir/builder.cpp (hunk 4 above)
Comment (jacobhinkle) [reply csrc/ir/builder.cpp:404]: I'll do that.

**Comment by naoyam** [comment]
> Question for you @zasdfgbnm regarding safety of these simplifcations: some of the other `SimplifyingIrBuilder` methods check for nullptrs, but not all do. When would we expect to have null inputs?

This class was originally meant to be a quick convenience class, and I did the nullptr handling just because it was convenient in some cases.

File: csrc/C++20/compare [hunk 6]
```diff
@@ -0,0 +1,40 @@
+// clang-format off
+/*
+ * SPDX-FileCopyrightText: Copyright (c) 2023-present NVIDIA CORPORATION & AFFILIATES.
+ * All rights reserved.
+ * SPDX-License-Identifier: BSD-3-Clause
+ */
+// clang-format on
+#pragma once
+
+namespace std {
```
Comment (zasdfgbnm) [inline csrc/C++20/compare:10]: Could you guard this with this macro?
https://github.com/NVIDIA/Fuser/blob/2abae36f572023a9c31200144997bd3e43a6b585/lib/dynamic_type/test/member.cpp#L76-L82

**Review by zasdfgbnm — APPROVED** [review]

File: csrc/C++20/compare (hunk 6 above)
Comment (jacobhinkle) [reply csrc/C++20/compare:10]: Done. Thanks. I also cleaned up the simplifications you mentioned earlier.

**Pull request merged**